### DISCUSSION TITLE
Infer attrs for implementing HOCs

### DISCUSSION THREAD

**Discussion by rudyxu1102** [vuejs/rfcs#479] [RFC Discussions]
- Start Date: 2023-01-01
- Target Major Version: 3.x
- Reference Issues:
  - https://github.com/vuejs/core/issues/3452
  - https://github.com/vuejs/core/issues/5423
  - https://github.com/vuejs/core/discussions/6528
  - https://github.com/vuejs/core/discussions/7546
  - https://github.com/vuejs/core/discussions/7981
  - https://github.com/vuejs/core/discussions/8264
  - https://github.com/vuejs/core/discussions/8508
  - https://github.com/vuejs/core/discussions/8620
  - https://github.com/vuejs/core/discussions/8902
  - https://github.com/vuejs/core/discussions/9272
- Implementation PR: 
  - https://github.com/vuejs/core/pull/7444
  - Volar

# Summary
Allowing to infer attrs by using `attrs` option of `defineComponent` or `defineCustomElement`. 
And in the `setup-script`,  passing generic type in the `defineAttrs<T>` will also infer `attrs` to `T`.

# Basic example

## Using `defineComponent`
Options Api
```tsx
const Comp = defineComponent({
  props: {
    foo: String
  },
  attrs: Object as AttrsType<{
    bar?: number
  }>,
  created() {
    this.$attrs.bar // number | undefined
  }
});

<Comp foo={'str'} bar={1} />;
```

Composition Api
```tsx
const Comp = defineComponent({
  props: {
    foo: {
      type: String,
      required: true
    }
  },
  attrs: Object as AttrsType<{
    bar?: number
  }>,
  setup(props, { attrs }) {
    props.foo; // string
    attrs.bar; // number | undefined
  }
});
<Comp foo={'str'} bar={1} />;
```
Functional Components
```tsx
const Comp = defineComponent(
  (props: { foo: string }, ctx) => {
    ctx.attrs.bar; // number | undefined
    return () => (
      <div>{props.foo}</div>
    )
  },
  {
    attrs: Object as AttrsType<{
      bar?: number
    }>
  }
);
<Comp foo={'str'} bar={1} />;
```

## Using `defineAttrs<T>` in `setup-script`

```vue
// MyImg.vue
<script setup lang="ts">
const attrs = defineAttrs<{ bar?: number }>();
</script>
```
<details open>
<summary>Complie Output: </summary>

```tsx
import { useAttrs as _useAttrs, defineComponent as _defineComponent } from 'vue'

export default /*#__PURE__*/_defineComponent({
  setup(__props, { expose: __expose }) {
  __expose();

      const attrs = _useAttrs()
      
return { attrs }
}

})
```

</details>

## Using `defineCustomElement`
```tsx
const Comp = defineComponent({
  props: {
    foo: String
  },
  attrs: Object as AttrsType<{
    bar?: number
  }>,
  created() {
    this.$attrs.bar // number | undefined
  }
});
```

# Motivation
This proposal is mainly to infer `attrs` using `defineComponent`.

When using typescript in Vue3, the fallthrough attributes is unable to be used. It's not appropriate obviously that only one can be chosen from `typescript` and `Fallthrough Attributes`. In most cases, we choose `typescript` and set attributes to `props` option instead of using the fallthrough attributes.

### 1. Wrapping a native HTML element in a new component, such as `img`. 

- Option Api
```tsx
import { defineComponent, type ImgHTMLAttributes, type AttrsType } from 'vue';

const MyImg = defineComponent({
    props: {
        foo: String
    },
    attrs: Object as AttrsType<ImgHTMLAttributes>,
    created() {
        this.$attrs.class // any
        this.$attrs.style // StyleValue | undefined
        this.$attrs.onError // ((payload: Event) => void) | undefined
        this.$attrs.src // string | undefined
    },
    render() {
        return <img {...this.$attrs} />
    }
});

<MyImg class={'str'} style={'str'} src={'https://xxx.com/aaa.jpg'} onError={(e) => {
  e; // Event
}}/>;
```
- Setup Script
```vue
// MyImg.vue
<script setup lang="ts">
import { type ImgHTMLAttributes } from 'vue';
const attrs = defineAttrs<ImgHTMLAttributes>();
</script>
```

### 2. Wrapping a component from UI library in a new component, such as `el-button` from `element-plus`. 

- Option Api
```tsx
import { defineComponent, type AttrsType } from 'vue';

const Comp = defineComponent({
    props: {
        foo: String
    },
    emits: {
        baz: (val: number) => true
    },
    render() {
        return <div>{this.foo}</div>
    }
});

const MyComp = defineComponent({
    props: {
        bar: Number
    },
    attrs: Object as AttrsType<typeof Child>,
    created() {
        this.$attrs.class // unknown
        this.$attrs.style // unknown
        this.$attrs.onBaz; // ((val: number) => any) | undefined
        this.$attrs.foo; // string | undefined
    },
    render() {
        return <Comp {...this.$attrs} />
    }
});

<MyComp class={'str'} style={'str'} bar={1} foo={'str'} onBaz={(val) => { 
    val; // number
}} />;
```

- Setup Script
```vue
// MyButton.vue
<script setup lang="ts">
import { Button } from 'element-plus';
const attrs = defineAttrs<typeof Button>();
</script>
```
# Unresolved questions
Naming suggestions or improvements on the API are welcome.

**Comment by cawa-93** [comment]
How it will work in functional components?

**Reply by rudyxu1102** [reply]
Here is [TS Playground](https://www.typescriptlang.org/play?jsx=1#code/JYWwDg9gTgLgBAbzgEwKYDNgDtUGELgQ5YwA0cMAnmKnAMqowCuY+JqAHvAL5zpQE4AcgBuTVAFoYAZwlMYwADbShAWABQoSLERwAFnF79Bo8UIDccDRqo04+cAEEYMKNLgBeRBrhwARgCGUABccFhMIH6oUD7+AQBeAPyh0q7YAOYa3NbqtrQOYACiIMAynt7qvgDGegFY6aihABQiAYopafUAlJ4AfHAiEMDI5lk5VUSpcACylAXlaJg4BUSoJE2xTQD6YAJg0qFI6BAQHVAZhuRVMByhDMysRDCcMAA8BcWl0uQFzq7SvR6Hn6CFivgA9OCwhEojFKr44BMsNIIIpUAA6RQQdJNa4cdEBFxudGBKBdMFwSFwVLnepwAA+cCYWEW2FQyApSJRaMx2NxNwJROkJIS5PUnIFqBKMCaQhqdQaQnIQgAjEIxRSoIwmFAsPpZchgCJ1bFsuoNeoufAJuByq9ZvNjhAPAhVUJeKSXSreODehogA)  about functional components.

**Comment by jods4** [comment]
I don't understand the motivation or the design.

`attrs` is supposed to be the catch-all for anything bound on a component that the component doesn't know about (i.e. `props`).
How/why would you type that? Aren't re-inventing props? Why is a well-known, typed attribute not a `prop`?

In motivation you say: 
> When using typescript in Vue3, the fallthrough attributes is unable to be used. 

As a Typescript user, I don't understand. I use TS and fall-through with no obvious issue, so I think it'd be worthwhile going into a little more details what the limitation is. Maybe even open an issue about it, there might be a more idiomatic way to fix the limitation you encounter without turning `attrs` into `props2`.

**Reply by cawa-93** [reply]
My point for this is intellisense for wrappers around html element. Lets say I have button-component that just add some default styles. 
```vue
<template>
  <button><slot/></button>
</template>

<style scoped> /* ... */ </style>
```
I want use Fallthrough Attributes here and I would really like to receive in my code editor all the hints about attributes, such as autocompletion, or messages that this attribute is used incorrectly.

**Reply by jods4** [reply]
I see, I have written several components that just enhance "native" tags in this way and have experienced the same.

It doesn't bother me that much for HTML controls because there's usually few tag-specific attributes, but it's much worse when you wrap a Vue component (as you said: to provide some contextually meaningful defaults).

The fallthrough itself works but you've lost all `props` typings from the inner component.

This can be worked-around by re-declaring all props and re-binding them (en masse using `v-bind`), but it's a tedious and brittle approach, as the `props` type must be locally declared in the same file and can't be imported from another file (currently a tooling limitation); so you must duplicate everything and keep it in sync when it changes -- esp. terrible if the component comes from third-party and has tons of props.

As much as I'd love to see a solution to this problem, I am not sure how this RFC solves it.
Sure, the implicit fall-through of `attrs` removes the need for `v-bind`, but that's IMHO by far not the worst issue here.
You would still need to re-declare all properties of the inner component.

For this specific goal, it'd be kind of nice if the tooling would have some smarts when a Component has `inheritAttrs` and a single inner component -> IDE provides some additional default props from underlying component.

**Reply by leopiccionia** [reply]
Yeah, something like `useAttrs<HTMLButtonElement>` (or like `useAttrs<Omit<HTMLButtonElement, keyof _ComponentProps>>`) would be cool. Better yet if it was automatically inferred.

**Reply by rudyxu1102** [reply]
> Yeah, something like `useAttrs<HTMLButtonElement>` (or like `useAttrs<Omit<HTMLButtonElement, keyof _ComponentProps>>`) would be cool. Better yet if it was automatically inferred.

`useAttrs<T>` is in the proposal.

**Reply by jods4** [reply]
Note that you're basically asking for Vue tooling to lift its single-file limitation.
That would be super cool but it's a big thing to ask as a sub-part of a larger RFC.

More details if it's unclear what I mean by that: 
To be useful, tooling needs to know what the type `HTMLButtonElement` is, which implies it sees the whole TS compilation (it comes from tslib). Likewise, it would need to see the props type of Vue components in other files.

AFAIK this is currently a limitation of `defineProps` in script setup, and it would apply similarly to a new `defineAttrs`.

**EDIT**: Or _maybe_ not. I _think_ the limitation applies to the Vue compiler, as `defineProps` is translated into a `props` definition object (it needs to know all the keys and the tool only looks at a single file without full TS context).
As long as the new macro does not impact view compilation (i.e. it's fully discarded), it should be ok. 
I _think_ Volar has access to the full TS compilation so IDE support should be ok.

**Reply by rudyxu1102** [reply]
> AFAIK this is currently a limitation of defineProps in script setup, and it would apply similarly to a new defineAttrs.

I agree with you. This Proposal still not resolves the limitation of `defineProps`. I think that it's related to unresolved questions of [script-setup-RFC](https://github.com/vuejs/rfcs/blob/master/active-rfcs/0040-script-setup.md#unresolved-questions). That would be cool if someone implement this feature.

**Comment by cawa-93** [comment]
Also, do not forget about the events. For example, a `<details>` element has not only a specific `open` attribute, but also a specific `toggle` event. IDE should propose it for `v-on`

**Reply by rudyxu1102** [reply]
Thank you for such positive feedback. It works on specific events.
I wrote a demo that wrapped `<image>` element. Please clone [this repository](https://github.com/rudy-xhd/vue-demo/tree/image-element) (branch: `image-element`) and run it.

**Reply by cawa-93** [reply]
> Thank you for so such positive feedback. It works on specific events. I wrote a demo that wrapped `<image>` element. Please clone [this repository](https://github.com/rudy-xhd/vue-demo/tree/image-element) and run it.

It's doesn't work in webstorm
![зображення](https://user-images.githubusercontent.com/1662812/210836758-13c267b6-ada9-472f-8c3f-cd7a098119d7.png)
![зображення](https://user-images.githubusercontent.com/1662812/210836789-64766844-8eae-498d-8d5f-9fdd20837c38.png)
🤷🏻‍♂️

**Reply by rudyxu1102** [reply]
It's related to https://github.com/johnsoncodehk/volar/issues/1240. It works fine in VS Code.

**Comment by rudyxu1102** [comment]
## Updated
Since the `slots` option is already supported, I think using the `attrs` option may be more in line with user habits,  instead of using the second parameter as stated in the RFC.

###  1. Option Component
```tsx
const Comp = defineComponent({
    slots: Object as SlotsType<{
        foo?: { data: string }
    }>,
    attrs: Object as AttrsType<{
        bar?: string
    }>,
    setup(props, { slots, attrs }) {
        console.log(attrs.bar)
        slots.foo?.({ data: 'a' })
    }
});
<Comp bar={"str"} />;
```
### 2. Functional Component
```tsx
const Comp = defineComponent(
  (props: { foo: string }, ctx) => {
    console.log(ctx.attrs.bar)
    return () => (
      <div>{props.foo}</div>
    )
  },
  {
    slots: Object as SlotsType<{
        baz?:  { data: string }
    }>,
    attrs: Object as AttrsType<{
      bar?: number
    }>
  }
);
<Comp bar={1} foo={"str"} />;
```
See https://github.com/vuejs/core/pull/7444

**Comment by rudyxu1102** [comment]
## Specific scenes
### 1. Wrapping a html element
```tsx
import { defineComponent, type ImgHTMLAttributes, type AttrsType } from 'vue';

const MyImg = defineComponent({
    props: {
        foo: String
    },
    attrs: Object as AttrsType<ImgHTMLAttributes>,
    created() {
        this.$attrs.class // any
        this.$attrs.style // StyleValue | undefined
    },
    render() {
        return <img {...this.$attrs} />
    }
});
<MyImg class={'str'} style={'str'} src={'str'} />;
```
### 2. Wrapping a component
```tsx
import { defineComponent, type AttrsType, type ExtractPropsAndEvents } from 'vue';

const Child = defineComponent({
    props: {
        foo: String
    },
    emits: {
        baz: (val: number) => true
    },
    render() {
        return <div>{this.foo}</div>
    }
});

const Comp = defineComponent({
    props: {
        bar: Number
    },
    attrs: Object as AttrsType<ExtractPropsAndEvents<typeof Child>>,
    created() {
        this.$attrs.class // unknown
        this.$attrs.style // unknown
    },
    render() {
        return <Child {...this.$attrs} />
    }
});

<Comp class={'str'} style={'str'} bar={1} foo={'str'} onBaz={(val) => { 
    val; // number
}} />;
```

**Reply by rudyxu1102** [reply]
## Update
I'm going to use `AttrsType` directly instead of `ExtractPropsAndEvents`, because using `ExtractPropsAndEvents` will confuse the user with `attrs` and `props`.

### 2. Wrapping a component
```tsx
import { defineComponent, type AttrsType } from 'vue';

const Child = defineComponent({
    props: {
        foo: String
    },
    emits: {
        baz: (val: number) => true
    },
    render() {
        return <div>{this.foo}</div>
    }
});

const Comp = defineComponent({
    props: {
        bar: Number
    },
    attrs: Object as AttrsType<typeof Child>,
    created() {
        this.$attrs.class // unknown
        this.$attrs.style // unknown
    },
    render() {
        return <Child {...this.$attrs} />
    }
});

<Comp class={'str'} style={'str'} bar={1} foo={'str'} onBaz={(val) => { 
    val; // number
}} />;
```

**Comment by rudyxu1102** [comment]
## Update
In order to be consistent with the usage habits of `defineSlots`, I plan to use `defineAttrs` instead of `useAttrs`.

**Reply by pikax** [reply]
Can you please update the discussing and the rendered when updating the proposal, instead of updating comments or the implementation

**Reply by rudyxu1102** [reply]
Updated.

**Comment by pikax** [comment]
I have a few questions that this RFC does not seem to be clear:

- What would be the difference between a `prop` and a `attribute`? 
- What would happen if the user sent the wrong type to an attribute? In props we do runtime validation.
- Do the plugins or tools need to make changes to support this RFC?
- Should attributes be implicitly optional? 
- How the user understands: 
```ts
// both 
const comp = defineComponent({  props: { foo: Number}, attrs: null as { bar: Number} })

// runtime error
<Comp foo={"1"}/> 
// not runtime error
<Comp bar={"1"}/>
```


--- 

I would argue if you're expecting the user to pass a specific prop, you should be controlling that through properties, since you're expecting those properties, since you even declaring them... 

Currently there's already ways to do what this SFC  is trying to add:
<img width="281" alt="image" src="https://github.com/vuejs/rfcs/assets/4620458/51f477b1-e19f-4b7b-850c-abe23832baa7">
Comp.vue
```html
<script setup lang="ts">
const props = defineProps<{
    foo: number
}>();
</script>
```
[playground](https://play.vuejs.org/#eNp9UlFPwjAQ/iuXxggkBGJ4GwOjhgd9UKI+9mWMA4db27QdYMj+u3edwIjTLEt63/dd77vrHcSdMYNtiSISsUttZjw49KWZSpUVRlsPD7owsLK6gM5gyEFnLFU8rMUko8BjYfLEI0UAcZ2g9USKGylg2EAXiW2gZzxqlzfgy8wWQXSpiIcnU6IvvEu1WmXrwcZpRa0e+AopUrojy9G+GJ9p5aSIIDDMJXmud08B87bE/hFPPzD9bME3bs+YFHOLDu0WpThxPrFr9DU9e3vGPZ1PZKGXZU7qf8hXdDov2WMtuy/Vkmw3dMHtY3iwTK3f3WzvUbljU2yUlVXQS0HvzeP7q/Wz3dFgFPKkqmiKnNO2K5Anak2z93TNVFqpaNjOg7HaOJjAEleZwjlH8YFprkIPF4EqiwW1QVA17fbGfDgt1rGed1Tu1ypKUe8i25EiZPKP+6CjikmZ/+gTB13/ZVCvQtyDaziAwl23Ry3DVXDJJ9qf26MlqPgT1TeeQAM7)

For the usage inside of the component, you can always cast your attributes:
```ts
// I would agree supporting T on `defineAttrs` instead.
const attrs =defineAttrs() as { bar?: number }
```

**Reply by rudyxu1102** [reply]
Good point and thanks for your review. Actually, I also use the implementation you mentioned above in my project, like the following code.

[TS Playground](https://www.typescriptlang.org/play?jsx=1#code/JYWwDg9gTgLgBAbzgEwKYDNgDtUGELgQ5bwC+c6UBcA5AG4CuqNA3ALABQokscMAnmFSI4+QsRgAFKmADOccpWr0mAWgDGBSBNUChqgBaoANkKizWnTnuEBRAB4xUWZGO3OYAHjcAaUQeBjZAA+OABeOBwAdzgACgBKcNCETjg4ABIwGVkALlEtIg9pCDlvLVCAMnzxIuzvAKDgzlIrDk0sWXg3cJQMbDwCiViUjjS0rJLcxFSxtPQICDyAZRgobABzGbTSHy24VBBgGCmR2bSAIwBDAC882LpL4zysBhBz1ChEsNDVpj2dvZQZxoKAJaajM5AmAMKBYOCeZDAOjBBAwAKyAB08wgpE8AHpEcj-s14uwOJx2p04ABZfjdCJoTA4NyFEjDPYTOR5U5nK5QPIAOVe7yg-12ELS6iBlycyDBPNmlPgSCIACEbn5sQoemjgJj0jLVvJLvIWRJiqUbBB0NUwE0JQpxbMgS4PvK9mkoTC4WVwIgMQHdfrDeZyHj7bMWhxSKTWkqafwAIIMGAQAAqglQri0OQcThcZo8nitNtpvj4metttCETLWjJnE8tOTqYzQmzfvUxhNsjCCBonSgNHInX4xlQfYHq2HcD5fYAjIoFpPBzO1Tc+-dHl9knA9g9jCw4Hi8ZFhR9mmHgiwgA)

This is a simple and efficient way, but you need to cast type **twice** at least and this is a bit difficult to understand for typescript newbies. I have no problem with this implementation, but I think it needs to be added in the document.

**Reply by rudyxu1102** [reply]
> What would be the difference between a prop and a attribute?

`props` can be thought of as attributes of the current component, while `attrs` can be thought of as attributes of the child component.

> What would happen if the user sent the wrong type to an attribute? In props we do runtime validation.

- If you wrap a vue component, there is still a runtime validation. [Vue SFC Playground](https://play.vuejs.org/#eNqlU8Fu2zAM/RVBKNYWaGKsvWVugK3oYTt0xbqjLo7NJOpkSZDkJIPhfx8p2U7SLsGA+WSKT+TTe2TLP1s73TTAZzz3pZM2MA+hsXOhZW2NC+zB1JYtnanZ5TSjgOCXQudZwiMSgwC1VUUAjBjL0x1j7gVfg1JGcLYoHEZb41SFUYa4PBsv8RsefGn0Uq6mr95oZNNSIcFLrCQVuO82SKO94DMWM5QrsPL2WzwLroGb4bxcQ/nrL+evfkdngj878OA2IPiYC4VbQUjpx5cn2OH/mKxN1ShEn0n+AG9UQxwT7EujK6R9gItsv0ZNpV799I+7ANoPjyKihOwiXnDUmEQ89fQ93bvpXbwndIcqDv4c2KkKvULlAxbYm9qyCpZSxx5Ggw6s6z0mcz+NuPDbAoJH2LMz1h9iJ2RQTE0IO0G7LTh/UOJhLVW1HyCK+glCw30/X/dv+Vz1Tx3L+/H19MUyKUyKMWaJ2hEIB/Aopo9IzthLcOjBPoPqDaW6a6FhF5kjpaJRPcPCx7tmmcIPqIqG7dU1dmAXfe9jmfIBT1znKBp2Obc064/ztiXSrOvyDKO0SlG+zWQhdYU2XhQhOLIyT1K+26NR4P8egd6e2P+kP+9Vx0U/ofpTUy9wJf5V9WTxWcVuSTFsmBS7fSNG9weNc5xx).
- If you wrap an html element, then as you said there is no runtime validation. [Vue SFC Playground](https://play.vuejs.org/#eNp9UttO4zAQ/RXLQgtI3UYr3roBiV0hAeIm4NEvIZmmBmds2ZO2KMq/M3ZDyk31SzKXM+fMpZOnzk2XLciZzEPptSMRgFp3olA3znoS168XTS3m3jZif5olKwL2FebZBsG5bBA0zhQEbAmRDyhrj5VcgDFWSVF6G4L1utbI3pX1pmJvxoA8G9FyIimUFue6nj4HiyysixWVLG3jtAF/60hbDErORIrEWMEMq8vkI9/C5N1fLqB8+cH/HNbRp+SdhwB+CUqOMSp8DbQJnz3cwJr/x2Bjq9Zw9o7gPQRr2qhxk/avxYplf8hLai/SeDXWj+FsTYDhvakoNGb2KV9JHvb/Ha1v5R5NjxJOYc9THDf1YbWmwJpHT1xhu+BOVDDXmEgsAtJE0KsDwfDzx+urUyKvn1qCIPrhDOL6/yrkLQUSESaOv9Y4GPQ5b10Y1cbHN/HJji/yzcQDE2G9jXAj6cNz6A8VwjrJZaKiNQNvERLWzjfmL24GYXVwyAxib+D+qQ8uvet6F39Oui4qFX2fZ2wlr+aLXv5+0ljxDPcKrsdz/Ha//RtnShmz)

> Do the plugins or tools need to make changes to support this RFC?

Volar needs support the `defineAttrs` macro.

> Should attributes be implicitly optional?

The attributes extracted from a Vue component or an html element have been processed to determine whether they are optional.

> How the user understands:
> // both 
> const comp = defineComponent({  props: { foo: Number}, attrs: null as { bar: Number} })
> 
> // runtime error
> <Comp foo={"1"}/> 
> // not runtime error
> <Comp bar={"1"}/>

I think jsx verification and runtime verification can be separated. In fact, some components now have the situation you mentioned above, e.g. [Element Plus Playground](https://element-plus.run/#eyJzcmMvQXBwLnZ1ZSI6IjxzY3JpcHQgc2V0dXAgbGFuZz1cInRzXCI+XG5pbXBvcnQgeyByZWYgfSBmcm9tICd2dWUnXG5jb25zdCBtc2cgPSByZWYoJ0hlbGxvIFdvcmxkIScpXG48L3NjcmlwdD5cblxuPHRlbXBsYXRlPlxuICA8aDE+e3sgbXNnIH19PC9oMT5cbiAgPGVsLWlucHV0IHYtbW9kZWw9XCJtc2dcIiA6YXV0b3NpemU9XCJ7IGFhYTogJzEnfVwiIC8+XG48L3RlbXBsYXRlPlxuIiwiaW1wb3J0LW1hcC5qc29uIjoie1xuICBcImltcG9ydHNcIjoge31cbn0iLCJ0c2NvbmZpZy5qc29uIjoie1xuICBcImNvbXBpbGVyT3B0aW9uc1wiOiB7XG4gICAgXCJhbGxvd0pzXCI6IHRydWUsXG4gICAgXCJjaGVja0pzXCI6IHRydWUsXG4gICAgXCJqc3hcIjogXCJwcmVzZXJ2ZVwiLFxuICAgIFwidGFyZ2V0XCI6IFwiRVNOZXh0XCIsXG4gICAgXCJtb2R1bGVcIjogXCJFU05leHRcIixcbiAgICBcIm1vZHVsZVJlc29sdXRpb25cIjogXCJCdW5kbGVyXCIsXG4gICAgXCJhbGxvd0ltcG9ydGluZ1RzRXh0ZW5zaW9uc1wiOiB0cnVlLFxuICAgIFwidHlwZXNcIjogW1wiZWxlbWVudC1wbHVzL2dsb2JhbC5kLnRzXCJdXG4gIH0sXG4gIFwidnVlQ29tcGlsZXJPcHRpb25zXCI6IHtcbiAgICBcInRhcmdldFwiOiAzLjNcbiAgfVxufVxuIiwiX28iOnt9fQ==). This example has strict jsx validation but no strict runtime validation.

**Comment by jods4** [comment]
Here's a take on this issue with minimal impact on Vue core and its type system.
What if Volar / vue-tsc / vscode-typescript-vue-plugin did the following trick with types:

**If the component has `inheritAttrs: true` then these tools should merge the Props types of its root component into its own Props type -- but for the Props generic type parameter only, not as real runtime-defined props.**

No runtime nor component metadata change: child component props are still undefined attrs on the wrapper, that are simply inherited.

No change to the complex Vue Component type either, as we piggy-back on the existing Props generic argument. 
Indeed attrs that are inherited into props can be handled "as if" they were identical props on the parent.

With this, during compilation we'd get errors for incorrect types, like for real props.
In IDE we'd get completion, like for real props.
And this could be automatic thanks to Volar plugins, without even any code change.

**Reply by pikax** [reply]
I'll create an RFC and a PoC https://github.com/vuejs/core-vapor/issues/4#issuecomment-1851473301, I think that would be the way to tackle this sort of issue for HoC

**Comment by teleskop150750** [comment]
It would be useful to define types for them

**Reply by jods4** [reply]
Attributes are reactive, too?

**Comment by jods4** [comment]
To anyone interested: I might have a workaround that provides strong `inheritAttrs` typing.
Hopefully this will be native to Vue language tools one day, until then you can look here: https://github.com/vuejs/language-tools/issues/4882#issuecomment-2375092076

**Reply by teleskop150750** [reply]
Isn't wrapper typing solved in [4103](https://github.com/vuejs/language-tools/pull/4103) ?

And yet I wanted to type attributes independently of child components.

**Reply by jods4** [reply]
Yeah, looks like so! Thanks for pointing it out, I missed that update!

Although... does it really work?
I have set `vueCompilerOptions: { fallthroughAttributes: true }` in my tsconfig, also tried with `strictTemplates: true` (which creates tons of unjustified errors BTW) and yet I don't get the completion I expect.

Component `screen-form` wraps `ui-form`, without any fallback option declared (so attrs fallback and that works).
Inside `screen-form` I get completion when I type `:diff` on ui-form.
In other files where I use `screen-form` I get no completion for `:diff`, isn't that what's supposed to work or did I miss something?

**Reply by jods4** [reply]
Nevermind the not-working part, I figured it out: it doesn't work when the wrapper component is generic. I'll see if there's an issue or open one.

**Comment by nayakunin** [comment]
I've recently encountered the same issue. I wanted to do something like this in Vue
```jsx
interface Props extends React.ComponentProps<"div"> {
  something: string;
}

export const Component = ({ something, ...props }: Props) => {
  return <div {...props}>{something}</div>;
};
```

Here is the same component in Vue.

```ts
<script setup lang="ts">
import type { HTMLAttributes } from "vue";

interface Props extends /* @vue-ignore */ HTMLAttributes {
  something: string;
}

defineProps<Props>();
</script>

<template>
  <div>
    {{ something }}
  </div>
</template>
```

The reason why I've started looking into this is that I wanted to enable `strictTemplates`, as I wanted to have full typescript coverage of my components. I immediately encountered an issue while doing that, because in any case where a fallthrough attribute is passed to a component, typescript doesn't know what it is and throws an error.

It works fine and in the consumers you get full type safety not only on regular properties, but also for callbacks. It correctly defines which types the events have in callbacks like `@click`.

> I have set vueCompilerOptions: { fallthroughAttributes: true } in my tsconfig, also tried with strictTemplates: true (which creates tons of unjustified errors BTW) and yet I don't get the completion I expect.

@jods4 why do you think these errors are unjustifiable? I've added an example bellow, where this would be quite beneficial, as it would catch an actual mistake.

I've also tested this approach without the `extends HTMLAttributes`, but with `fallthroughAttributes: true` and what I observed is that the inference only works on variable attributes, like an id, but doesn't work on callbacks. Do you know if that's something that's being worked on?

Here is an example why I believe the ability to define the types on attributes (or properly infer) is quite important. This code doesn't trigger any typescript errors, as the `event` has a type of `any`. 
```
<Component @click="(event) => event.idk()" />
```
 
There is an example from `LinusBorg` in another [discussion](https://github.com/orgs/vuejs/discussions/10262#discussioncomment-8354644) showcasing how this might work given the current limitations, but it seems quite cumbersome.

**Reply by jods4** [reply]
Sorry, it was almost one year ago, I don't have all examples at hand anymore.

It creates useful typings and errors, but I recall that in my project it also produced lots of false negatives (i.e. things that work and shouldn't be reported as errors). At the time, this + the lack of generic support + the DX perf hit made me stick to my hand-rolled solution (see my comment above if you're curious).

**Comment by GrantGryczan** [comment]
I'm confused what the purpose of this is considering there's already a `tsconfig.json` option to do this automatically under `vueCompilerOptions`: [`fallthroughAttributes`](https://github.com/vuejs/language-tools/wiki/Vue-Compiler-Options#fallthroughattributes-v210)

Defining attribute types manually is not type-safe, and the motivation described by the RFC seems to be completely covered by the `fallthroughAttributes` option already. If there's some reason the `fallthroughAttributes` option doesn't cover the same use case as this, I think that should be explicitly mentioned in the RFC.

EDIT: Actually, this would still be very useful if Vue language tools had an option to enforce that the correct fallthrough attribute types are explicitly defined by `defineAttrs`. That would address the scaling performance issue with `fallthroughAttributes`. This really should be documented in the RFC.

**Comment by KazimirPodolski** [comment]
@rudyxu1102 Sir is there a chance to get this back on track?

Making wrappers is a pain in Vue and wrappers is what you routinely do with UI frameworks in everyday development. This change would greatly help with the pain.
